src/contrib/cuda.jl

```diff
@@ -0,0 +1,27 @@
+import .CUDA
+
+CUDA.allowscalar(false)
+
+function refresh(
+    rng::Union{AbstractRNG, AbstractVector{<:AbstractRNG}},
+    ::FullMomentumRefreshment,
+    h::Hamiltonian,
+    z::PhasePoint{T}
+) where {T<:CUDA.CuArray}
+    r = CUDA.CuArray{Float32, 2}(undef, size(h.metric)...)
+    CUDA.CURAND.randn!(r)
+    return phasepoint(h, z.θ, r)
+end
+
+function mh_accept_ratio(
+    rng::Union{AbstractRNG, AbstractVector{<:AbstractRNG}}, Horiginal::TA, Hproposal::TA,
+) where {T<:AbstractFloat, TA<:CUDA.CuArray{<:T}}
+    α = min.(one(T), exp.(Horiginal .- Hproposal))
+    # NOTE: There is a chance that sharing the RNG over multiple
+    #       chains for accepting / rejecting might couple
+    #       the chains. We need to revisit this more rigirously 
+    #       in the future. See discussions at 
+    #       https://github.com/TuringLang/AdvancedHMC.jl/pull/166#pullrequestreview-367216534
+    accept = TA(rand(rng, T, length(Horiginal))) .< α
```

Comment:
Two comments here.
1. It seems that we can also make `TA(rand())` to directly sample on GPU instead of sample on CPU and move to GPU. Should give us some extra perfermance gain but I did miss it in my gist version.
2. It would be nice to merge this whole function with the CPU version somehow. This is not urgent if there is not clear way but worth put a note in the comment as TODO.

Reply:
Looking at the [CUDA documentation](https://juliagpu.gitlab.io/CUDA.jl/usage/array/#Random-numbers) it looks like it is possible to pass in a CUDA random number generator which directly samples on the GPU. I will look into whether that way we can avoid having special GPU functions for `mh_accept_ratio` and `refresh`.

Reply:
I used the `randn` trick as an initial solution. My comment was in response to your second point above but that discussion can be left for a separate PR!